### DISCUSSION TITLE
Create case-insensitive Literal type hint

### DISCUSSION THREAD

**Discussion by tvdboom** [beartype/beartype#285] [Q&A] [answered]
Hi,

I am trying to create a `Literal` type hint that's case-insensitive, since I don't want to bother the user with case-sensitive nonsense.
I have created this class, that works...
```python
import typing
from beartype.typing import Annotated
from beartype.vale import Is

class Literal(typing.Literal):
    def __class_getitem__(cls, X: tuple[str]) -> Annotated[str, Is]:
        return Annotated[str, Is[lambda x: x.lower() in X]]
```
... but mypy complains. For example, when having a function like this
```python
def func(item: Literal["a", "b"] = "a"): ...

```
it raises this error: `error: Incompatible default for argument "item" (default has type "str", argument has type "Literal")`. So mypy doesn't recognize my class `Literal` as a `str`, while this works for the original `typing.Literal`.

Any idea how to fix this? Thanks

**Comment by leycec** [comment, marked as answer]
**Fascinating...** and exceedingly clever, too. That's just what we've all come to expect from @tvdboom.

Surprisingly, this now is beginning to seem like more of a mypy issue rather than a @beartype issue. I blame them! Mypy is bad and should feel bad.

Specifically, mypy fails to extend the same semantics applied to standard `typing` classes to subclasses of those standard `typing` classes. If mypy accepts a function annotated by `typing.Literal[A, B]` for any valid literals `A` and `B`, then mypy should also accept the same function annotated by `YourLiteral[A, B]` for any subclass `class YourLiteral(typing.Literal):`.

But... it doesn't. So, that's a mypy bug. Right? I think? I hope? Please let this not be @beartype's responsibility. *Odin, hear our cry!* :ghost:

**Reply by tvdboom** [reply]
Fair enough. Looking at the issue tracker of mypy, I think my changes are slim of getting any help there... I'll try to find a workaround while I wait for an answer (https://github.com/python/mypy/issues/16172). Thanks anyway

**Reply by leycec** [reply]
*...yeah.* Pretty much. Static type-checkers really have *no* incentive to resolve issues specific to runtime type-checkers like @beartype. We are the competition. They'd prefer us to just go away.

The only workaround I can think of is the obvious (and very, very dumb) workaround: litter your code with filthy `# type: ignore[{whatevah}]` comments every time you want to use your homebrew `Literal` type. Of course, this is yet another reason why static type-checkers are bad for everyone's health: these inscrutable, highly opinionated, and mostly **non-configurable** black boxes are wrong about almost everything.

"Mostly non-configurable." That's the problem. Sure, you could *try* to [configure mypy through mypy-specific plugins](https://mypy.readthedocs.io/en/stable/extending_mypy.html#extending-mypy-using-plugins). That would probably work. But `pylance`, which is even more popular than mypy, fails to provide a similar plugin API. There's currently no standard general-purpose off-the solution to customize static type-checking behaviour.

Thus, I sigh. :face_exhaling:
